src/fspec-fromdir.c

```diff
@@ -0,0 +1,113 @@
+#define _XOPEN_SOURCE 500
+#include <ftw.h>
+#include <limits.h>
+#include <unistd.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdint.h>
+#include <errno.h>
+
+int root_owned = 0;
+int absolute = 0;
+char *prefix = "";
+
+static void
+check(int ok) {
+    if (!ok) {
+        if (errno != 0) {
+            fprintf(stderr, "failed - %s\n", strerror(errno));
+        } else {
+            fputs("failed\n", stderr);
+        }
+        exit(1);
+    }
+}
+
+static const char *
+filetype(int tflag)
+{
+    if (tflag == FTW_F)
+        return "reg";
+    if (tflag == FTW_D)
+        return "dir";
+    if (tflag == FTW_SL)
+        return "sym";
+    fprintf(stderr, "unknown file type flags '%d'\n", tflag);
+    exit(1);
+}
+
+static int
+printfspec(const char *fpath, const struct stat *sb,
+            int tflag, struct FTW *ftwbuf)
+{
+    int len;
+    char pathbuf[PATH_MAX];
+
+    if (strcmp(fpath, ".") == 0)
+        return 0;
+
+    check(printf("%s%s\n", prefix, fpath+2) > 0);
+    check(printf("type=%s\n", filetype(tflag)) > 0);
+    check(printf("mode=%o\n", sb->st_mode & 0777 /* XXX masking too much */ ) > 0);
```

Comment:
You can use `sb->st_mode & ~S_IFMT` here.

Also, use `%04o` in the format string so that it is always 4 characters long.